exporters/otlp/trace/src/main/java/io/opentelemetry/exporter/otlp/trace/MarshalerInputStream.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.exporter.otlp.trace;
+
+import com.google.common.io.ByteStreams;
+import com.google.protobuf.CodedOutputStream;
+import io.grpc.Drainable;
+import io.grpc.KnownLength;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import javax.annotation.Nullable;
+
+// Adapated from gRPC ProtoInputStream but using our Marshaller
+// https://github.com/grpc/grpc-java/blob/2c2ebaebd5a93acec92fbd2708faac582db99371/protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoInputStream.java
+final class MarshalerInputStream extends InputStream implements Drainable, KnownLength {
+  @Nullable private Marshaler message;
+  @Nullable private ByteArrayInputStream partial;
+
+  MarshalerInputStream(Marshaler message) {
+    this.message = message;
+  }
+
+  @Override
+  public int drainTo(OutputStream target) throws IOException {
+    int written;
+    if (message != null) {
+      written = message.getSerializedSize();
+      CodedOutputStream cos = CodedOutputStream.newInstance(target);
+      message.writeTo(cos);
+      cos.flush();
```

Comment:
For better or worse we have more test coverage since the days of #2890 - the Armeria-based tests that actually exercise Netty showed this missing flush from the old PR. I point it out since it means even after the random SpanData generation for verifying marshaling behavior, there are still points that could cause issues. But I think with the tests in this repo and eventually the java instrumentation smoke tests, we'll be fine and have to make the final leap.